witnesslint.py

```diff
@@ -0,0 +1,1128 @@
+                # All other expected children have already been handled and removed
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    child.sourceline,
+                    "Graph element has unexpected child of type '%s'",
+                    child.tag,
+                )
+
+    def handle_graphml_elem(self, graphml_elem):
+        if len(graphml_elem.attrib) > 0:
+            self.log_with_position(
+                LOGLEVELS["warning"],
+                graphml_elem.sourceline,
+                "Expected graphml element " "to have no attributes",
+            )
+        if None in graphml_elem.nsmap:
+            if graphml_elem.nsmap[None] != "http://graphml.graphdrawing.org/xmlns":
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    graphml_elem.sourceline,
+                    "Unexpected default namespace: %s",
+                    graphml_elem.nsmap[None],
+                )
+        else:
+            self.log_with_position(
+                LOGLEVELS["warning"],
+                graphml_elem.sourceline,
+                "Missing default namespace",
+            )
+        if "xsi" in graphml_elem.nsmap:
+            if graphml_elem.nsmap["xsi"] != "http://www.w3.org/2001/XMLSchema-instance":
+                self.log_with_position(
+                    LOGLEVELS["warning"],
+                    graphml_elem.sourceline,
+                    "Expected 'xsi' to be namespace prefix for "
+                    "'http://www.w3.org/2001/XMLSchema-instance'",
+                )
+        else:
+            self.log_with_position(
+                LOGLEVELS["warning"],
+                graphml_elem.sourceline,
+                "Missing xml schema namespace "
+                "or namespace prefix is not called 'xsi'",
+            )
+        for child in graphml_elem:
+            # All expected children have already been handled and removed
+            self.log_with_position(
+                LOGLEVELS["warning"],
+                graphml_elem.sourceline,
+                "Graphml element has unexpected child of type '%s'",
+                child.tag,
+            )
+
+    def final_checks(self):
+        """
+        Performs checks that cannot be done before the whole witness has been traversed
+        because elements may appear in almost arbitrary order.
+        """
+        for key in self.used_keys.difference(set(self.defined_keys)):
+            if key in COMMON_KEYS:
+                # Already handled for other keys
+                self.log(
+                    LOGLEVELS["warning"], "Key '%s' has been used but not defined", key
+                )
+        for key in set(self.defined_keys).difference(self.used_keys):
+            self.log(
+                LOGLEVELS["info"],
+                "Unnecessary definition of key '%s', key has never been used",
+                key,
+            )
+        if self.witness_type is None:
```

Comment:
But should this really be replaced? Imho this way is more readable and in some cases it makes a difference whether `x is None`  or `not x` is used, so I would leave it everywhere as-is for consistency and to prevent accidentally altering the semantics.